src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,414 @@
+    auto & nodes_info = cluster_info.nodes_info;
+
+    if (std::find(node_uuids.begin(), node_uuids.end(), current_node_name) == node_uuids.end())
+    {
+        LOG_ERROR(log, "Can't find current node in cluster '{}', will register again", cluster_info.name);
+        registerInZk(zk, cluster_info);
+        nodes_info.clear();
+        return false;
+    }
+
+    if (!needUpdate(node_uuids, nodes_info))
+    {
+        LOG_TRACE(log, "No update required for cluster '{}'", cluster_info.name);
+        return true;
+    }
+
+    nodes_info = getNodes(zk, cluster_info.zk_root, node_uuids);
+    if (nodes_info.empty())
+    {
+        LOG_WARNING(log, "Can't get nodes info for '{}'", cluster_info.name);
+        return false;
+    }
+
+    int current_version;
+    getNodeNames(zk, cluster_info.zk_root, cluster_info.name, &current_version, true);
+
+    if (current_version != start_version)
+    {
+        LOG_TRACE(log, "Cluster '{}' configuration changed during update", cluster_info.name);
+        nodes_info.clear();
+        return false;
+    }
+
+    LOG_TRACE(log, "Updating system.clusters record for '{}' with {} nodes", cluster_info.name, cluster_info.nodes_info.size());
+
+    auto cluster = makeCluster(cluster_info);
+    context->setCluster(cluster_info.name, cluster);
+    return true;
+}
+
+bool ClusterDiscovery::updateCluster(const String & cluster_name)
+{
+    auto cluster_info = clusters_info.find(cluster_name);
+    if (cluster_info == clusters_info.end())
+    {
+        LOG_ERROR(log, "Unknown cluster '{}'", cluster_name);
+        return false;
+    }
+    return updateCluster(cluster_info->second);
+}
+
+void ClusterDiscovery::registerInZk(zkutil::ZooKeeperPtr & zk, ClusterInfo & info)
+{
+    LOG_DEBUG(log, "Registering current node {} in cluster {}", current_node_name, info.name);
+
+    String node_path = getShardsListPath(info.zk_root) / current_node_name;
+    zk->createAncestors(node_path);
+
+    zk->createOrUpdate(node_path, info.current_node.serialize(), zkutil::CreateMode::Ephemeral);
+    LOG_DEBUG(log, "Current node {} registered in cluster {}", current_node_name, info.name);
+}
+
+void ClusterDiscovery::start()
+{
+    if (clusters_info.empty())
+    {
+        LOG_DEBUG(log, "No defined clusters for discovery");
+        return;
+    }
+    LOG_TRACE(log, "Starting working thread");
+
+    auto zk = context->getZooKeeper();
+    for (auto & [_, info] : clusters_info)
+    {
+        registerInZk(zk, info);
+        if (!updateCluster(info))
+        {
+            LOG_WARNING(log, "Error on updating cluster '{}', will retry", info.name);
+            clusters_to_update->set(info.name);
+        }
+    }
+
+    main_thread = ThreadFromGlobalPool([this]
+    {
+        try
+        {
+            runMainThread();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, "Caught exception in cluster discovery runMainThread");
+        }
```

Comment:
There may be a plenty of exceptions. E.g. session can expire.
I think we should take a new zk session and retry in a loop.
Other exceptions may probably be ignored. However, it is also not cool if we get memory limit exception and just stop discovery forever.

Reply:
done